### DISCUSSION TITLE
Decouple from WPF

### DISCUSSION THREAD

**Discussion by ShemJM** [canton7/Stylet#348] [General]
Is it possible to decouple this from WPF?  It'd be good to be able to create ViewModels that can be used in WPF, WinUI, Xamarin etc and keep using Stylet which I find to be the easiest MVVM framework to work with. 

Also are you a member of the C# discord server? Lots of talk about MVVM and Stylet happening there. https://discord.gg/csharp

**Comment by canton7** [comment]
A lot of it is inherently tied to WPF. Thr only things which aren't particularly are the Screen/Conductor stuff and the ioc container (although Screen.Parent is a UIElement, which is WPF). Actions, the window manager, the VM-first support (s:View.Model etc) are necessarily tied to WPF.

I have done a cut-down version for Xamarin.Forms for a couple of work projects: you can implement the VM-first stuff pretty well, the window manager needs to be completely rethought, and actions to ICommand work but actions with events don't.

Basically, a lot of what Stylet does necessarily involves working very closely with WPF, so it's not just a case of writing some cross-framework code once and it magically working everywhere: supporting a new framework means rewriting (and testing) about half of Stylet.

That's not to say that can't be done of course, but I work mainly with WPF (professionally and personally), so supporting other frameworks would be a large time investment which I wouldn't benefit from. So there's been little motivation to do the work "just because". 

I've got my eye on MAUI - I'm interested to see how that turns out, and whether there's a need for something like Stylet. If they've designed it well, you won't need Stylet to fill in the gaps, after all! 

I am in the C# Discord, but I don't hang out there often and when I do it's mainly in roslyn/lowlevel, so I hadn't noticed Stylet was being mentioned, thanks! It's a shame you can't set up keyword notifications - feel free to @ me if there's something I can help with.

**Reply by canton7** [reply]
Having said that, ValidatingModelBase is all built around INotifyDataErrorInfo, which while not WPF specific, isn't used by WinUI. I can't find any information on how MAUI input validation...

**Reply by ShemJM** [reply]
Okay, that makes sense. It does seem like it would be more work than it's worth. I mostly work with WPF at the moment too, just if in the future we move to WinUI/Maui it'd mean rewriting a lot of the VMs. Probably not such a bad thing as they are terribly written 😅 

Oh nice, I see you there now. Most WPF/MVVM talk is done in the GUI channel.

**Reply by WillBellJr** [reply]
I'm in the same boat - I absolute LOVE WPF (and hence Stylet), but likewise I've been watching all the new Maui conferences on YouTube, and admittedly, I've always wanted to be able to write Android apps using C#. (I was watching over the weekend a YT video on how now, you can create **Blazor** apps that run natively (you can access the hardware), on PC, Android etc. (basically a _web viewer_ is used behind the scene for UI rendering - they had the same app running on Windows, Mac and Android accessing network availability queries on all three - YUMMY!)

I'm thinking, no, **hoping** this will be the next wave of "WPF" development, that I can learn and enjoy (I hated dealing w/Javascript when writing ASP.NET MVC/Core apps here at work - slows me way down!  I've grabbed a Udemy Maui course just to familiarize myself in preparation...)

I've had so much success using WPF & Stylet, that I don't want to give it up - we also use DevExpress controls here, and I prefer Stylet over the MVVM framework DevEx created.
(I'm sitting here now working with the DevEx "Wizard Control" trying to get my ViewModels to receive Activation events when switching between their Wizard Pages - I just find their MVVM _hoakey_ by comparison!)

-Will

**Comment by canton7** [comment]
I'm having a big internal debate on whether to base any "Common" library on the MVVM toolkit, or continue to use `PropertyChangedBase`/`ValidatingModelBase`.

On the one hand, the question of "Stylet or MVVM Toolkit" comes up a lot on Discord, and taking the MVVM Toolkit as a dependency would stop that conversation quite nicely. On the other hand, it means that the current Stylet wouldn't be able to depend on the new Common library, which would cause all sorts of problems.

Unless we start a proliferation of Stylet.Common, Stylet.Common.MvvmToolkit, Stylet.Wpf, Stylet.Wpf.MvvmToolkit, etc...

**Reply by alexhelms** [reply]
Is there any update on splitting out `ValidatingModelBase` to use CommunityToolkit? I'm starting a new project and would love to be able to use the source generator features in CommunityToolkit to save on lots of boilerplate.

**Reply by alexhelms** [reply]
Adding on to this after some more digging -- I saw you had a branch from 2 years ago (probably when you made this comment 😆) experimenting with splitting off `ValidatingModelBase`. Do you think the time is right to finish that? I'm willing to contribute to bring it across the finish line.
